### PR TITLE
build improvements

### PR DESCRIPTION
**Pull request by yamt**
TODO:

* ~~maybe `imagebuild.sh` should use `make container` as well~~ test this

### CODE REVIEW

**Comment by marun** [comment]
@yamt Is this ready for merge or do you want to complete that TODO first?

**Comment by yamt** [comment]
@marun either way works for me

File: scripts/deploy-federation.sh [hunk 1]
```diff
@@ -172,10 +172,10 @@ fi
...
 # Build federation binaries and image
 if [[ ! "${USE_LATEST}" ]]; then
-  base_dir="$(cd "$(dirname "$0")/.." ; pwd)"
-  dockerfile_dir="${base_dir}/images/federation-v2"
-  CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -o "${dockerfile_dir}"/hyperfed "${base_dir}"/cmd/hyperfed/main.go
-  docker build ${dockerfile_dir} -t "${IMAGE_NAME}"
+  saved_dir="$(pwd)"
+  cd "$(dirname "$0")/.."
+  make container IMAGE_NAME=${IMAGE_NAME}
+  cd "${saved_dir}"
```
Comment (shashidharatd) [inline scripts/deploy-federation.sh]: Cant we just use `cd -` here?

File: Makefile [hunk 2]
```diff
@@ -84,25 +76,40 @@ fmt:
...
 
 $(BIN_DIR):
 	mkdir $(BIN_DIR)
 
-hyperfed: $(BIN_DIR)
-	$(DOCKER_BUILD) '$(BUILD_HYPERFED)'
+$(HYPERFED_TARGET)-linux: $(BIN_DIR)
+	$(DOCKER_BUILD) '$(TARGET_ENV_LINUX) $(GO_BUILDCMD) -o $(HYPERFED_TARGET)-linux cmd/hyperfed/main.go'
+
+$(HYPERFED_TARGET)-darwin: $(BIN_DIR)
```
Comment (shashidharatd) [inline Makefile]: (suggestion) Maybe we could have reduced duplicate lines if the Makefile target is something like `$(HYPERFED_TARGET)-$(HOST_OS): `

Comment (yamt) [reply Makefile]: i made some changes along the line.
https://github.com/kubernetes-sigs/federation-v2/pull/684/commits/5046bf44b63abdb21dce502c43f4797be1a9cd86

File: scripts/deploy-federation.sh (hunk 1 above)
Comment (yamt) [reply scripts/deploy-federation.sh]: done

File: Makefile [hunk 3]
```diff
@@ -84,25 +79,36 @@ fmt:
...
+COMMANDS := $(HYPERFED_TARGET) $(CONTROLLER_TARGET) $(KUBEFED2_TARGET)
+OSES := linux darwin
+ALL_BINS :=
+
+define OS_template
+$(1)-$(2): bindir
+	$(DOCKER_BUILD) '$(TARGET_ENV_$(2)) $(GO_BUILDCMD) -o $(1)-$(2) cmd/$(3)/main.go'
+ALL_BINS := $(ALL_BINS) $(1)-$(2)
+endef
+$(foreach cmd, $(COMMANDS), $(foreach os, $(OSES), $(eval $(call OS_template, $(cmd),$(os),$(notdir $(cmd))))))
```
Comment (shashidharatd) [inline Makefile]: It does not make sense to build binaries for all OSes as a default as it is unnecessary compilation time overhead. So please only build for the HOST OS and should be the default.
For release build, only kubefed2 is the client side binary( so it makes sense to build & release kubefed2 for all OS). the container image will be based on Linux for now.

File: Makefile [hunk 4]
```diff
@@ -37,27 +41,18 @@ endif
...
                       -X github.com/kubernetes-sigs/federation-v2/pkg/version.gitCommit=$(GIT_HASH) \
                       -X github.com/kubernetes-sigs/federation-v2/pkg/version.gitTreeState=$(GIT_TREESTATE) \
                       -X github.com/kubernetes-sigs/federation-v2/pkg/version.buildDate=$(BUILDDATE)"
-BUILDCMD_HYPERFED = CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -o $(HYPERFED_TARGET) $(VERBOSE_FLAG) $(LDFLAG_OPTIONS)
-BUILDCMD_CONTROLLER = CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -o $(CONTROLLER_TARGET) $(VERBOSE_FLAG) $(LDFLAG_OPTIONS)
-BUILDCMD_KUBEFED2 = go build -o $(KUBEFED2_TARGET) $(VERBOSE_FLAG) $(LDFLAG_OPTIONS)
-
-BUILDCMD_CONTROLLER_NATIVE = CGO_ENABLED=0 go build -o $(CONTROLLER_TARGET_NATIVE) $(VERBOSE_FLAG) $(LDFLAG_OPTIONS)
+TARGET_ENV_linux = GOARCH=amd64 GOOS=linux
+TARGET_ENV_darwin = GOARCH=amd64 GOOS=darwin
```
Comment (shashidharatd) [inline Makefile]: we could avoid these 2 variables, if we just use GOOS=$(HOST_OS)

File: Makefile (hunk 3 above)
Comment (yamt) [reply Makefile]: it only builds for the host os if you do eg. "make controller"

File: Makefile (hunk 4 above)
Comment (yamt) [reply Makefile]: done

**Comment by shashidharatd** [comment]
Thanks @yamt,
/lgtm

**Comment by poothia** [comment]
lgtm

**Pull request merged**